File: datascripts/lib/canonical.ts
// The canonical form of a Package Delta, written to byte-match
// `crates/lyracore-package-delta/src/canonical.rs`.
//
// Byte equality is the interop contract, not a nicety: the importer re-canonicalizes every artifact
// it reads and the Module hashes what it applies, so an artifact this library writes and the same
// artifact round-tripped through Rust must be the same bytes.
//
// The rules, all of them, in the same order the Rust writer states them:
//
//  * No whitespace anywhere, and no trailing newline.
//  * Object members appear in a fixed declared order; `fields` members appear sorted by name.
//  * Claims appear sorted by table, then spell, then effect index.
//  * Integers are plain decimal.
//  * An unsigned 64-bit value is a decimal string with no sign, no padding and no separators.
//  * A float is the shortest decimal that reads back as the same `f32`, always with a decimal
//    point.
//  * A string escapes only what JSON requires, using the short escape where one exists.

import { SPELL_EFFECT_TABLE, SPELL_TABLE, type FieldType } from "./schema.ts";

/// The artifact version this library writes.
export const DELTA_VERSION = 1;

/// One claimed column value, already narrowed to the column's type.
export interface FieldValue {
  readonly type: FieldType;
  readonly value: number | bigint | boolean | string;
}

/// The row a claim names, typed by its table.
export type PrimaryKey =
  | { readonly kind: "spell"; readonly spellId: number }
  | {
      readonly kind: "spellEffect";
      readonly spellId: number;
      readonly effectIndex: number;
    };

/// One Package's claim on one row. The table follows from the key, so a claim cannot name a table
/// its key does not describe.
export interface Claim {
  readonly key: PrimaryKey;
  readonly operation: "insert" | "update";
  readonly fields: Map<string, FieldValue>;
}

/// The durable table a key's row lives in.
export function tableOf(key: PrimaryKey): string {
  return key.kind === "spell" ? SPELL_TABLE : SPELL_EFFECT_TABLE;
}

/// The row as a conflict report writes it, which is also how a duplicate-claim refusal names it.
export function describeKey(key: PrimaryKey): string {
  return key.kind === "spell"
    ? `{spell_id=${key.spellId}}`
    : `{spell_id=${key.spellId}, effect_index=${key.effectIndex}}`;
}

/// The whole artifact in canonical form. `claims` is sorted here rather than by the caller, so the
/// order is a property of the writer and not of the order an author happened to write things in.
export function writeDelta(
  packageId: string,
  sourceHash: string,
  claims: Claim[],
): string {
  const ordered = [...claims].sort(compareKeys);
  const parts = ordered.map(writeClaim).join(",");
  return `{"version":${DELTA_VERSION},"package":${writeString(packageId)},"source_hash":${writeString(sourceHash)},"claims":[${parts}]}`;
}

/// The canonical claim order: every `game_spell` row before every `game_spell_effect` row, then by
/// spell, then by effect index. Mirrors the derived `Ord` on the Rust `PrimaryKey`.
export function compareKeys(a: Claim, b: Claim): number {
  const rank = (claim: Claim) => (claim.key.kind === "spell" ? 0 : 1);
  if (rank(a) !== rank(b)) return rank(a) - rank(b);
  if (a.key.spellId !== b.key.spellId) return a.key.spellId - b.key.spellId;
  if (a.key.kind === "spellEffect" && b.key.kind === "spellEffect") {
    return a.key.effectIndex - b.key.effectIndex;
  }
  return 0;
}

function writeClaim(claim: Claim): string {
  const names = [...claim.fields.keys()].sort();
  const fields = names
    .map((name) => `${writeString(name)}:${writeValue(claim.fields.get(name)!)}`)
    .join(",");
  return `{"table":${writeString(tableOf(claim.key))},"key":${writeKey(claim.key)},"operation":${writeString(claim.operation)},"fields":{${fields}}}`;
}

function writeKey(key: PrimaryKey): string {
  return key.kind === "spell"
    ? `{"spell_id":${key.spellId}}`
    : `{"spell_id":${key.spellId},"effect_index":${key.effectIndex}}`;
}

function writeValue(value: FieldValue): string {
  return `{"type":${writeString(value.type)},"value":${scalarLiteral(value)}}`;
}

/// One claimed value in its canonical JSON spelling, without the surrounding type tag.
export function scalarLiteral(value: FieldValue): string {
  switch (value.type) {
    case "u64":
      // A JSON number loses precision above 2^53, so an unsigned 64-bit value travels as a string.
      return `"${value.value}"`;
    case "f32":
      return f32Literal(value.value as number);
    case "bool":
      return value.value ? "true" : "false";
    case "string":
      return writeString(value.value as string);
    default:
      return String(value.value);
  }
}

/// A float as Rust's `f32` `Display` writes it, plus the decimal point Rust's writer appends: the
/// shortest decimal that reads back as the same `f32`, never in exponent form.
export function f32Literal(input: number): string {
  const text = shortestF32(Math.fround(input));
  return text.includes(".") ? text : `${text}.0`;
}

/// The shortest decimal that round-trips through `f32`. `toPrecision` is correctly rounded, so the
/// first precision that survives the round trip is the same digit string Rust's shortest-repr writer
/// produces.
function shortestF32(value: number): string {
  if (!Number.isFinite(value)) {
    throw new RangeError(`${value} is not a finite f32`);
  }
  if (Object.is(value, -0)) return "-0";
  for (let precision = 1; precision <= 9; precision++) {
    const candidate = value.toPrecision(precision);
    if (Math.fround(Number(candidate)) === value) return plainDecimal(candidate);
  }
  return plainDecimal(value.toPrecision(9));
}

/// The same value with the exponent multiplied out and no padding zeros. Rust's float `Display`
/// never uses exponent form, so neither may this.
function plainDecimal(text: string): string {
  let sign = "";
  let body = text;
  if (body.startsWith("-")) {
    sign = "-";
    body = body.slice(1);
  }

  let exponent = 0;
  const marker = body.indexOf("e");
  if (marker >= 0) {
    exponent = Number(body.slice(marker + 1));
    body = body.slice(0, marker);
  }

  const dot = body.indexOf(".");
  let digits = dot >= 0 ? body.slice(0, dot) + body.slice(dot + 1) : body;
  let pointAt = (dot >= 0 ? dot : body.length) + exponent;

  while (pointAt <= 0) {
    digits = `0${digits}`;
    pointAt += 1;
  }
  while (pointAt > digits.length) {
    digits = `${digits}0`;
  }

  const whole = digits.slice(0, pointAt).replace(/^0+(?=\d)/, "");
  const fraction = digits.slice(pointAt).replace(/0+$/, "");
  return sign + whole + (fraction ? `.${fraction}` : "");
}

/// A JSON string escaping only what JSON requires, with the short escape where one exists — the
/// exact set the Rust writer emits.
function writeString(text: string): string {
  return JSON.stringify(text);
}

File: crates/lyracore-package-delta/src/canonical.rs
//! The canonical form of a Package Delta.
//!
//! Two artifacts that say the same thing must produce the same bytes, so that a source hash, a diff
//! and a signature all mean what they look like they mean. The writer is hand-built rather than
//! derived, because the byte stability is this crate's promise and must not move under a library's
//! formatting default.
//!
//! The rules, all of them:
//!
//!  * No whitespace anywhere, and no trailing newline.
//!  * Object members appear in a fixed declared order; `fields` members appear sorted by name.
//!  * Claims appear sorted by table, then spell, then effect index.
//!  * Integers are plain decimal. An integer column written as `100.0` or `1e2` is refused at the
//!    parse rather than rounded here, so there is only ever one spelling to write.
//!  * An unsigned 64-bit value is a decimal string with no sign, no padding and no separators.
//!  * A float is the shortest decimal that reads back as the same `f32`, always with a decimal
//!    point, so `1`, `1.0` and `1.00` all become `1.0`.
//!  * A string escapes only what JSON requires, using the short escape where one exists.

use core::fmt;

use crate::delta::{Claim, PackageDelta, PrimaryKey, DELTA_VERSION};
use crate::schema::FieldValue;

pub(crate) fn write_delta(delta: &PackageDelta) -> String {
    let mut out = String::new();
    out.push_str("{\"version\":");
    out.push_str(&DELTA_VERSION.to_string());
    out.push_str(",\"package\":");
    write_json_string(&mut out, delta.package().as_str());
    out.push_str(",\"source_hash\":");
    write_json_string(&mut out, delta.source_hash().as_str());
    out.push_str(",\"claims\":[");
    for (index, claim) in delta.claims().iter().enumerate() {
        if index > 0 {
            out.push(',');
        }
        write_claim(&mut out, claim);
    }
    out.push_str("]}");
    out
}

fn write_claim(out: &mut String, claim: &Claim) {
    out.push_str("{\"table\":");
    write_json_string(out, claim.table().as_str());
    out.push_str(",\"key\":");
    write_key(out, claim.key());
    out.push_str(",\"operation\":");
    write_json_string(out, claim.operation().as_str());
    out.push_str(",\"fields\":{");
    for (index, (name, value)) in claim.fields().iter().enumerate() {
        if index > 0 {
            out.push(',');
        }
        write_json_string(out, name);
        out.push(':');
        write_value(out, value);
    }
    out.push_str("}}");
}

// One arm per key shape, and no wildcard: the canonical member order IS this crate's promise, so a
// key that arrives without stating its order must fail the build rather than fall into a fallback.
// That is what makes this function long; splitting it would trade the guarantee for the line count.
#[allow(clippy::too_many_lines)]
fn write_key(out: &mut String, key: PrimaryKey) {
    match key {
        PrimaryKey::Spell { spell_id }
        | PrimaryKey::SpellChain { spell_id }
        | PrimaryKey::SpellProcEvent { spell_id } => {
            write_members(out, &[("spell_id", spell_id.into())]);
        }
        PrimaryKey::SpellEffect {
            spell_id,
            effect_index,
        } => write_members(
            out,
            &[
                ("spell_id", spell_id.into()),
                ("effect_index", effect_index.into()),
            ],
        ),
        PrimaryKey::Item { entry }
        | PrimaryKey::Quest { entry }
        | PrimaryKey::GossipMenu { entry }
        | PrimaryKey::CreatureTemplate { entry }
        | PrimaryKey::GameobjectTemplate { entry }
        | PrimaryKey::GameobjectTrap { entry } => write_members(out, &[("entry", entry.into())]),
        PrimaryKey::QuestText { quest_entry } => {
            write_members(out, &[("quest_entry", quest_entry.into())]);
        }
        PrimaryKey::QuestObjective {
            quest_entry,
            obj_index,
        }
        | PrimaryKey::QuestCastObjective {
            quest_entry,
            obj_index,
        } => write_members(
            out,
            &[
                ("quest_entry", quest_entry.into()),
                ("obj_index", obj_index.into()),
            ],
        ),
        PrimaryKey::QuestRewardItem {
            quest_entry,
            item_entry,
        } => write_members(
            out,
            &[
                ("quest_entry", quest_entry.into()),
                ("item_entry", item_entry.into()),
            ],
        ),
        PrimaryKey::QuestRewardChoice {
            quest_entry,
            choice_index,
        } => write_members(
            out,
            &[
                ("quest_entry", quest_entry.into()),
                ("choice_index", choice_index.into()),
            ],
        ),
        PrimaryKey::PickpocketLoot { id }
        | PrimaryKey::GameobjectLoot { id }
        | PrimaryKey::SkinningLoot { id }
        | PrimaryKey::FishingLoot { id }
        | PrimaryKey::CreatureSpell { id }
        | PrimaryKey::TrainerSpell { id }
        | PrimaryKey::NpcTextSlot { id }
        | PrimaryKey::CreateinfoSpell { id }
        | PrimaryKey::SpellLearn { id }
        | PrimaryKey::QuestEventRequirement { id } => write_members(out, &[("id", id)]),
        PrimaryKey::CreatureAiBroadcastText { id } | PrimaryKey::CreatureAiSummon { id } => {
            write_members(out, &[("id", id.into())]);
        }
        PrimaryKey::CreatureCast { creature_entry } => {
            write_members(out, &[("creature_entry", creature_entry.into())]);
        }
        PrimaryKey::GossipMenuProfile { menu_id } => {
            write_members(out, &[("menu_id", menu_id.into())]);
        }
        PrimaryKey::GossipMenuProfileOption { row_id } | PrimaryKey::GossipOption { row_id } => {
            write_members(out, &[("row_id", row_id.into())]);
        }
        PrimaryKey::NpcText { text_id } => write_members(out, &[("text_id", text_id.into())]),
        PrimaryKey::ClassLevelStats { class, level } => {
            write_members(out, &[("class", class.into()), ("level", level.into())]);
        }
        PrimaryKey::LevelStats { race, class, level } => write_members(
            out,
            &[
                ("race", race.into()),
                ("class", class.into()),
                ("level", level.into()),
            ],
        ),
        PrimaryKey::StartPosition { race, class } => {
            write_members(out, &[("race", race.into()), ("class", class.into())]);
        }
        PrimaryKey::GraveyardZone { row_id } | PrimaryKey::CreateinfoAction { row_id } => {
            write_members(out, &[("row_id", row_id)]);
        }
        PrimaryKey::AreatriggerTeleport { trigger_id } => {
            write_members(out, &[("trigger_id", trigger_id.into())]);
        }
        PrimaryKey::CreatureSpawn {
            map_id,
            entry,
            spawn_id,
        } => write_members(
            out,
            &[
                ("map_id", map_id.into()),
                ("entry", entry.into()),
                ("spawn_id", spawn_id.into()),
            ],
        ),
        PrimaryKey::GameobjectSpawn { map_id, spawn_id } => write_members(
            out,
            &[("map_id", map_id.into()), ("spawn_id", spawn_id.into())],
        ),
    }
}

/// A key object, its members in the order the key declares them. That order is part of the
/// canonical form: two artifacts that name the same row have to hash the same, so it is fixed here
/// rather than left to whatever a generator emitted.
fn write_members(out: &mut String, members: &[(&str, u64)]) {
    out.push('{');
    for (index, (name, value)) in members.iter().enumerate() {
        if index > 0 {
            out.push(',');
        }
        write_json_string(out, name);
        out.push(':');
        out.push_str(&value.to_string());
    }
    out.push('}');
}

fn write_value(out: &mut String, value: &FieldValue) {
    out.push_str("{\"type\":");
    write_json_string(out, value.field_type().as_str());
    out.push_str(",\"value\":");
    out.push_str(&scalar_literal(value));
    out.push('}');
}

/// One claimed value in its canonical JSON spelling, without the surrounding type tag. Shared with
/// `FieldValue`'s `Display`, so a conflict report quotes a value exactly as the artifact writes it.
pub(crate) fn scalar_literal(value: &FieldValue) -> String {
    let mut out = String::new();
    match value {
        FieldValue::U8(n) => out.push_str(&n.to_string()),
        FieldValue::U16(n) => out.push_str(&n.to_string()),
        FieldValue::U32(n) => out.push_str(&n.to_string()),
        FieldValue::U64(n) => {
            out.push('"');
            out.push_str(&n.to_string());
            out.push('"');
        }
        FieldValue::I32(n) => out.push_str(&n.to_string()),
        FieldValue::F32(n) => write_f32(&mut out, *n),
        FieldValue::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        FieldValue::Str(s) => write_json_string(&mut out, s),
    }
    out
}

/// `f32` `Display` already gives the shortest decimal that reads back as the same value and never
/// uses exponent form. The only thing missing is a decimal point on a whole number, which keeps a
/// float column visually and textually distinct from an integer one.
fn write_f32(out: &mut String, value: f32) {
    let text = value.to_string();
    out.push_str(&text);
    if !text.contains('.') {
        out.push_str(".0");
    }
}

pub(crate) fn write_json_string(out: &mut String, text: &str) {
    out.push('"');
    for ch in text.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{8}' => out.push_str("\\b"),
            '\u{c}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => {
                use fmt::Write as _;
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            c => out.push(c),
        }
    }
    out.push('"');
}
